substrate/runtime/contract/src/exec.rs

```diff
@@ -0,0 +1,257 @@
+// Copyright 2018 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate. If not, see <http://www.gnu.org/licenses/>.
+
+use super::{CodeOf, ContractAddressFor, Module, Trait};
+use account_db::{AccountDb, OverlayAccountDb};
+use gas::GasMeter;
+use vm;
+
+use rstd::prelude::*;
+use runtime_primitives::traits::Zero;
+use runtime_support::StorageMap;
+use staking;
+use system;
+
+pub struct CreateReceipt<T: Trait> {
+	pub address: T::AccountId,
+}
+
+pub struct CallReceipt {
+	pub return_data: Vec<u8>,
+}
+
+pub struct ExecutionContext<'a, T: Trait + 'a> {
+	// typically should be dest
+	pub self_account: T::AccountId,
+	pub overlay: OverlayAccountDb<'a, T>,
+	pub depth: usize,
+}
+
+impl<'a, T: Trait> ExecutionContext<'a, T> {
+	/// Make a call to the specified address.
+	pub fn call(
+		&mut self,
+		caller: T::AccountId,
+		dest: T::AccountId,
+		value: T::Balance,
+		gas_meter: &mut GasMeter<T>,
+		_data: Vec<u8>,
```

Comment:
If I understood this, we don't pass `_data` to `vm::execute` or `Ext` yet because it's unimplemented, right?

And just a note on quadratic attacks -- it looks like we decide to put callstack directly on program stack, so in the future, using `&[u8]` might be more efficient and lifetime should work.

Reply:
Yeah, exactly, it is untackled issue right now. And I didn't delve into it. 
And yes, in theory, we can directly read from/write to the caller memory.

But there is a complication with writing though. Maybe we will have to make a temporary copy. The reason for that is caller can't know the size of the return buffer before calling the callee, and thus, might want to allocate space after the return data size is known. This is an open research question right now.